exonum/src/blockchain/transaction.rs

```diff
@@ -0,0 +1,203 @@
+    /// #
+    /// use exonum::blockchain::Transaction;
+    /// use exonum::crypto::PublicKey;
+    /// use exonum::messages::Message;
+    /// # use exonum::blockchain::TransactionStatus;
+    /// # use exonum::storage::Fork;
+    ///
+    /// message! {
+    ///     struct MyTransaction {
+    ///         // Transaction definition...
+    /// #       const TYPE = 1;
+    /// #       const ID = 1;
+    /// #       const SIZE = 32;
+    ///         field public_key:    &PublicKey    [0 => 32]
+    ///     }
+    /// }
+    ///
+    /// impl Transaction for MyTransaction {
+    ///     fn verify(&self) -> bool {
+    ///         self.verify_signature(self.public_key())
+    ///     }
+    ///
+    ///     // Other methods...
+    ///     // ...
+    /// #   fn execute(&self, _: &mut Fork) -> TransactionStatus { TransactionStatus::Succeeded }
+    /// }
+    /// # fn main() {}
+    fn verify(&self) -> bool;
+
+    /// Takes the current blockchain state via `fork` and can modify it if certain conditions
+    /// are met.
+    ///
+    /// # Notes
+    ///
+    /// - When programming `execute`, you should perform state-related checks before any changes
+    /// to the state and return early if these checks fail.
+    /// - If the execute method of a transaction raises a `panic`, the changes made by the
+    /// transactions are discarded, but the transaction itself is still considered committed.
+    /// - A transaction execution status (see `ExecutionStatus` for the details) is stored in the
+    /// blockchain and can be accessed through api.
+    ///
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # #[macro_use] extern crate exonum;
+    /// #
+    /// use exonum::blockchain::{Transaction, TransactionStatus};
+    /// use exonum::crypto::PublicKey;
+    /// use exonum::storage::Fork;
+    ///
+    /// message! {
+    ///     struct MyTransaction {
+    ///         // Transaction definition...
+    /// #       const TYPE = 1;
+    /// #       const ID = 1;
+    /// #       const SIZE = 32;
+    /// #       field public_key:    &PublicKey    [0 => 32]
+    ///     }
+    /// }
+    ///
+    /// impl Transaction for MyTransaction {
+    ///     fn execute(&self, fork: &mut Fork) -> TransactionStatus {
+    ///         // Read and/or write into storage.
+    ///         // ...
+    ///
+    ///         // Returns execution status.
+    ///         TransactionStatus::Succeeded
+    ///     }
+    ///
+    ///     // Other methods...
+    ///     // ...
+    /// #   fn verify(&self) -> bool { true }
+    /// }
+    /// # fn main() {}
+    fn execute(&self, fork: &mut Fork) -> TransactionStatus;
+}
+
+/// Execution status of the transaction.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub enum TransactionStatus {
+    /// Successful transaction execution.
+    Succeeded,
+    /// Panic occurred during transaction execution. This status should not be used explicitly.
+    Panic,
```

Comment:
Naming seems slightly inconsistent. Should it be `Paniced` to match with `Failed`? If we are going to change this, then perhaps go with all nouns?

```
pub enum TransactionStatus {
   Success,
   Panic,
   Failure,
   Custom(u8),
}
```